File: CPlusPlus/Problem 004/Problem004.cpp
/*
Project Euler
Problem 4
"Find the largest palindrome made from the product of two 3-digit numbers."
*/

#include <iostream>

long reverseNum(long num)
{
	long t = 0;
	while (num > 0)
	{
		t *= 10;
		t += num % 10;
		num /= 10;
	}
	return t;
}

bool isPalindrome(long num)
{
	 return (reverseNum(num) == num);
}

int main (void)
{
	long largest = 0;
	long prod = 0;
	for (int a = 100; a < 1000; ++a)
	{
		for (int b = 100; b < 1000; ++b)
		{
			prod = a * b;
			if (isPalindrome(prod) && (prod > largest))
			{
				largest = prod;
			}
		}
	}

	std::cout << largest << std::endl;
	return 0;
}

File: Scala/src/main/scala/euler/Problem004.scala
package euler

// Project Euler
// Problem 4
//
// A palindromic number reads the same both ways. The largest palindrome made from the product
// of two 2-digit numbers is 9009 = 91  99.
//
// Find the largest palindrome made from the product of two 3-digit numbers.

object Problem004 {
	def reverseString(str: String) = {
		str.foldLeft("")((a, e)=> e+a)
	}

	def isPalindromic(n: Int) = {
		n.toString() == reverseString(n.toString())
	}

	def nums(a: Int, b: Int): Stream[Int] = {
		if (b < 100)
			if (a < 100) return Stream.Empty
			else return Stream.cons(a*b, nums(a-1, 999));
		else return Stream.cons(a*b, nums(a, b-1));
	}

	def main(args: Array[String]) = {
		println(nums(999, 999).filter(x => isPalindromic(x)).max)
	}
}